pipeline_branching.md

```diff
@@ -0,0 +1,159 @@
+
+An empty `branch` field is equivalent of a value of `"master"`.
+
+Different branches of the same pipeline are effectively different pipelines. That means:
+- They output to different branches, with names that match their branch names.
+- You can delete, update, pause, etc. one without pausing the other.
+- They can take different inputs, have different resource requests, etc.
+- They run in different pods, and their success / failure is independent.
+
+There are a few ways in which they're not quite totally distinct:
+- They consolidate in list pipeline, this one's mostly cosmetic, I'm
+imagining something that looks like this:
+```
+NAME     BRANCHES     INPUT    CREATED    STATE / LAST JOB  DESCRIPTION
+pipeline dev, staging input:/* 6 days ago running / success
+```
+- They share datums for deduplication, if you want a dev version to
+reprocess, you need to create it with `--reprocess`.
+- They can't take each other as input. Implementation wise there's no
+particular reason for this, but it seems very weird and I can't think of any
+reason you'd want to do this that there isn't a better way of accomplishing.
+
+### A few examples of what syntax would look like for this:
+Creating a dev branch:
+
+```
+extract pipeline pipeline >pipeline.json
+# modify pipeline.json, setting Pipeline.Branch to "dev"
+create pipeline -f pipeline.json
+```
+
+List jobs from dev branch:
+
+```
+list job -p pipeline@dev
+```
+
+Get files from dev branch (all other file commands work similarly):
+
+```
+get file pipeline@dev:/file
+```
+
+Promote dev branch to master
+
+```
+# modify pipeline.json, setting Pipeline.Branch to ""
+create pipeline -f pipeline.json
+```
+
+## Downstream Pipelines:
+The above changes are fairly straightforward, both in terms of how we
+would implement them and in terms of explaining them to users. Downstream
+pipelines is where it starts to get complicated. Downstream pipelines are
+important to consider because when you're iterating on a pipeline, you
+care that the results of your new version can be consumed by the rest of
+the DAG downstream. For example, suppose I have the pipelines: A, B and C
+with a DAG like this:
+
+```
+A -> B -> C
+```
+
+Initially, each pipeline has only a `master` version, then we create
+`A@dev` this should create branches of `B` and `C` also called, `dev`
+which process `A@dev` and `B@dev` respectively. So now the picture looks
+like this:
+
+```
+A -> B -> C
+A@dev -> B@dev -> C@dev
+```
+
+B@dev and C@dev are identical to B and C, except that they process the
+`dev` branch of their inputs, rather than `master`.
+
+Adding a new pipeline works similarly, if someone comes and creates `D`
+then we will automatically create `D@dev` which is identical to `D`,
+except that it takes `C@dev` instead of `C` as input. The picture would
+now look like this:
+
+```
+A -> B -> C -> D
+A@dev -> B@dev -> C@dev -> D@dev
+```
+
+This means that there is a special relationship between the `master`
```

Comment:
I thought provenance and subvenance would be sufficient to define the relationship. For the dev branch it would point to corresponding @dev. It would be good to put an example to show when provenance/subvenance relationship diverge from the parent-child.

Reply:
It's better to think parent child relationship as orthogonal to the prov subv relationship. This is the picture I have in my mind:

```
+------------+           +------------+
|            |           |            |
|   Repo A   +---PROV---->   Repo B   |
|            |           |            |
+-----+------+           +-----+------+
      |                        |
      |PARENT                  |PARENT
      |                        |
+-----v------+           +-----v------+
|            |           |            |
|   Repo C   +---PROV---->   Repo D   |
|            |           |            |
+------------+           +------------+
```

So you have the relationship that your Provenance is your parent's provenance's child, which is a bit of a mouthful, but gives us the relationship we want. If it was all provenance relationships we wouldn't be able to distinguish which ones we should inherit provenance, but not actually have as provenance, from our provenance.